packages/table/src/interactions/selectable.tsx

```diff
@@ -110,142 +124,92 @@ export class DragSelectable extends React.Component<IDragSelectableProps, {}> {
+        const {
+            allowMultipleSelection,
+            focusedCell,
+            locateClick,
+            locateDrag,
+            selectedRegions,
+            selectedRegionTransform,
+        } = this.props;
+
+        let region = allowMultipleSelection
+            ? locateDrag(event, coords, /* returnEndOnly? */ this.didExpandSelectionOnActivate)
+            : locateClick(event);
+
+        if (!Regions.isValid(region)) {
             return;
+        } else if (selectedRegionTransform != null) {
+            region = selectedRegionTransform(region, event, coords);
         }
+
+        const nextSelectedRegions = this.didExpandSelectionOnActivate
+            ? this.expandSelectedRegions(selectedRegions, region, focusedCell)
+            : Regions.update(selectedRegions, region);
+
         this.maybeInvokeSelectionCallback(nextSelectedRegions);
 
-        if (!this.props.allowMultipleSelection) {
-            // have the focused cell follow the selected region
-            const mostRecentRegion = nextSelectedRegions[nextSelectedRegions.length - 1];
-            const focusCellCoordinates = Regions.getFocusCellCoordinatesFromRegion(mostRecentRegion);
-            const fullFocusCellCoordinates: IFocusedCellCoordinates = {
-                col: focusCellCoordinates.col,
-                focusSelectionIndex: nextSelectedRegions.length - 1,
-                row: focusCellCoordinates.row,
-            };
-            this.props.onFocus(fullFocusCellCoordinates);
+        if (!allowMultipleSelection) {
+            // move the focused cell with the selected region
+            const lastIndex = nextSelectedRegions.length - 1;
+            const mostRecentRegion = nextSelectedRegions[lastIndex];
+            this.invokeOnFocusCallbackForRegion(mostRecentRegion, lastIndex);
         }
     }
 
-    private handleDragEnd = (event: MouseEvent, coords: ICoordinateData) => {
-        const nextSelectedRegions = this.getDragSelectedRegions(event, coords);
-        if (nextSelectedRegions == null) {
-            return;
-        }
-        this.maybeInvokeSelectionCallback(nextSelectedRegions);
-        BlueprintUtils.safeInvoke(this.props.onSelectionEnd, nextSelectedRegions);
+    private handleDragEnd = () => {
         this.finishInteraction();
     }
 
-    private handleClick = (event: MouseEvent) => {
-        if (this.shouldIgnoreMouseDown(event)) {
-            return false;
-        }
-
-        let region = this.props.locateClick(event);
-        const { selectedRegions } = this.props;
-
-        if (!Regions.isValid(region)) {
-            this.maybeInvokeSelectionCallback([]);
-            BlueprintUtils.safeInvoke(this.props.onSelectionEnd, []);
-            return false;
-        }
-
-        if (this.props.selectedRegionTransform != null) {
-            region = this.props.selectedRegionTransform(region, event);
-        }
+    private handleClick = () => {
+        this.finishInteraction();
+    }
```

Comment:
note that both are simple now. thought about passing `this.finishInteraction` directly in [`this.getDraggableProps`](https://github.com/palantir/blueprint/pull/1524/files#diff-6f85c4f864b85874d8deec61bac44cb0R117), but it felt clearer to have separate `handleDragEnd` and `handleClick` functions.